### PR TITLE
feat(*): Add automerge bot, generate token script, and  markdown for automerge action

### LINKED ISSUE #2393: Investigate using a bot to automerge PRs
**Please describe the Improvement and/or Feature Request**
There has been interest in having PRs automerged by a bot to avoid manually having to rebase and merge changes. 

I have looked into https://github.com/pascalgn/automerge-action and it works as advertised. The main limitation from using this bot is that a PR merged via the automerge action will not trigger workflows configured on push due to a Github limitation.

This issue is to investigate if we can leverage this bot while finding a meaningful alternative to not ignore a set of tests not run prior to merge.

**Scope (please mark with X where applicable)**
- New Functionality          [ ]
- Install                    [ ]
- SMI Traffic Access Policy  [ ]
- SMI Traffic Specs Policy   [ ]
- SMI Traffic Split Policy   [ ]
- Permissive Traffic Policy  [ ]
- Ingress                    [ ]
- Egress                     [ ]
- Envoy Control Plane        [ ]
- CLI Tool                   [ ]
- Metrics                    [ ]
- Certificate Management     [ ]
- Sidecar Injection          [ ]
- Logging                    [ ]
- Debugging                  [ ]
- Tests                      [ ]
- CI System                  [X]
- Project Release            [ ]

**Possible use cases**
The use case is to automerge PRs that are ready for merge.

### PR DESCRIPTION
**Pull request by shalier**
**Description**:
Adds automerge pull request feature and closes #2393 

**Affected area**:

- New Functionality      [ X]
- Documentation          [X ]
- Install                [ ]
- Control Plane          [ ]
- CLI Tool               [ ]
- Certificate Management [ ]
- Networking             [ ]
- Metrics                [ ]
- SMI Policy             [ ]
- Security               [ ]
- Tests                  [ ]
- CI System              [ ]
- Performance            [ ]
- Other                  [ ]

Please answer the following questions with yes/no.

- Does this change contain code from or inspired by another project? If so, did you notify the maintainers and provide attribution?
No

### CODE REVIEW

File: CONTRIBUTING.md
Comment (shashankram) [inline CONTRIBUTING.md:101]: This should be optional when someone wants to merge a pull request. By default review authors should be able to hold PRs as long as they want.

File: CONTRIBUTING.md
Comment (shashankram) [inline CONTRIBUTING.md:93]: Please retain this section, and provide additional instructions to support auto-merge via the bot.

**Review by shashankram — CHANGES_REQUESTED** [review]

File: docs/developer/automerge.md
Comment (nojnhuh) [inline docs/developer/automerge.md:8]: Are there any rules of thumb we should follow for when and when not to use the `automerge` label?

File: .github/workflows/automerge.yml
Comment (shashankram) [inline .github/workflows/automerge.yml:15]: Need to change this, because every time someone leaves a comment this PR it triggers the automerge GH action: https://github.com/openservicemesh/osm/actions/runs/549416822

Comment (shashankram) [reply .github/workflows/automerge.yml:15]: Seems like there is no better way, so we can keep this as is.

File: .github/workflows/automerge.yml
Comment (shashankram) [inline .github/workflows/automerge.yml:46]: 5 minutes seems a bit too long to retry a merge failure. If the automerge merge process fails, it would be better to know this as soon as possible to avoid further delays. Maybe setting it to 1 min (100000) could be more efficient.

File: docs/developer/automerge.md
Comment (shashankram) [reply docs/developer/automerge.md:8]: It is up to the maintainer at the moment to decide whether a specific PR can be automerged. I think a few rules should be documented regarding when the automerge label can be used. We haven't discussed this yet.

There are several cases where an automerge label will do more harm than good, especially when a PR needs more than just the minimum 2 reviews. Since it is not standard practice to put the `do-not-merge/hold` label for every PR that is pending reviews by default, we need to think through and outline in this doc on when a PR can safely use the automerge label.

File: .github/workflows/automerge.yml
Comment (shalier) [reply .github/workflows/automerge.yml:46]: @shashankram It's actually set to retry merging every 5 min because it'll keep checking in on when the status checks are finished. And then the e2e tests take on average like ~25min

**Comment by ritazh** [comment]
cc @sozercan to help take a look at the github action stuff

File: docs/developer/automerge.md
Comment (sozercan) [inline docs/developer/automerge.md:9]: How does this work with `paths-ignore` (like a docs change)? wouldn't all the tests be required in all the PRs?

File: docs/developer/automerge.md
Comment (sozercan) [inline docs/developer/automerge.md:10]: is it possible to keep only rebase/merge with main functionality but disable automerge to start with?

Comment (shalier) [reply docs/developer/automerge.md:10]: I think changing the `UPDATE_LABELS` to be like `rebase` would do it. Then PRs labelled with `rebase` should only rebase and   if the `automerge` label is added then it'll automerge, but let me test it out and I'll let you know!

File: docs/developer/automerge.md
Comment (shalier) [reply docs/developer/automerge.md:9]: you're right, all the status checks would have to be run for all PRs

File: docs/developer/automerge.md
Comment (shashankram) [reply docs/developer/automerge.md:10]: @shalier @sozercan why is there a need for additional labels like rebase?

Comment (shalier) [reply docs/developer/automerge.md:10]: Confirming that changing the `UPDATE_LABELS` to `rebase` will let the PR rebase without automerging

Comment (shalier) [reply docs/developer/automerge.md:10]: @shashankram the new `rebase` label would only be added if we wanted to only rebase PRs but not automerge them.

Comment (shashankram) [reply docs/developer/automerge.md:10]: Got it, that's a good feature to have!

File: docs/developer/automerge.md
Comment (sozercan) [reply docs/developer/automerge.md:9]: this is definitely a downside of this, it will eliminate any optional checks

File: docs/developer/automerge.md
Comment (sozercan) [reply docs/developer/automerge.md:10]: if what's needed is a bot that constantly rebases instead of automerges (which also eliminates optional checks), I think that's a better approach to start with.

File: docs/developer/automerge.md
Comment (shalier) [reply docs/developer/automerge.md:9]: @sozercan just added a script that will wait for the status checks to complete so that branch protection rules will not need to be enabled

File: docs/developer/automerge.md
Comment (shalier) [reply docs/developer/automerge.md]: @sozercan, @shashankram It turns out that setting the `UPDATE_LABELS` to `rebase` would mean that a PR will only rebase with the `rebase` label and that the `automerge` label would not rebase. And adding the `automerge` label in addition to the `rebase` label to `UPDATE_LABELS` wouldn't work as it then means in order to rebase you'd need both the `automerge` and `rebase` labels. For now, I've made it so that the `automerge` label will do both rebasing and automerging.

**Comment by shashankram** [comment]
Will automerge only happen if there is the automerge label?

**Review by shashankram — CHANGES_REQUESTED** [review]
Everything looks good, except there is one additional note that needs to be documented.
With the automerge bot we need to document that PRs should be labeled for automerge only once the author has verified their PR follows the merge requirements as described in https://github.com/openservicemesh/osm/blob/main/CONTRIBUTING.md#merging-pull-requests. 

Could you add this to the `Merging pull requests` section in the contributing guide as well.

File: CONTRIBUTING.md
Comment (nojnhuh) [inline CONTRIBUTING.md:93]: We should document who is expected to assign the label. Maybe it makes the most sense to follow the same pattern we have for manually merging PRs: If the author is a maintainer, then they assign the label. If the author is an external contributor, then any maintainer can assign the label.

**Review by shashankram — APPROVED** [review]

File: .github/workflows/automerge.yml
Comment (shalier) [reply .github/workflows/automerge.yml:20]: It's because the JWT will expire ten min after creation, so then if the workflow has to re-run it would need to wait for status checks and e2e's take ~15-20 min

**Review by nojnhuh — APPROVED** [review]

**Review by michelleN — APPROVED** [review]

**Pull request merged**
